File: src/main/java/AoC2022_12.java
import java.util.List;
import java.util.Set;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Stream;

import com.github.pareronia.aoc.CharGrid;
import com.github.pareronia.aoc.Grid.Cell;
import com.github.pareronia.aoc.graph.BFS;
import com.github.pareronia.aocd.Aocd;
import com.github.pareronia.aocd.Puzzle;

public class AoC2022_12 extends AoCBase {
    
    private final CharGrid grid;
    private final Cell end;
    
    private AoC2022_12(final List<String> input, final boolean debug) {
        super(debug);
        this.grid = CharGrid.from(input);
        trace(this.grid);
        this.end = this.grid.findAllMatching(c -> c == 'E')
                    .findFirst().orElseThrow();
        trace("end: " + this.end);
    }
    
    public static final AoC2022_12 create(final List<String> input) {
        return new AoC2022_12(input, false);
    }

    public static final AoC2022_12 createDebug(final List<String> input) {
        return new AoC2022_12(input, true);
    }
    
    private char getValue(final Cell cell) {
        final char ch = this.grid.getValue(cell);
        return switch (ch) {
            case 'S' -> 'a';
            case 'E' -> 'z';
            default -> ch;
        };
    }
    
    private int solve(final Set<Character> endPoints) {
        final Predicate<Cell> isEnd = cell ->
                endPoints.contains(this.grid.getValue(cell));
        final Function<Cell, Stream<Cell>> adjacent = cell ->
                this.grid.getCapitalNeighbours(cell)
                        .filter(n -> getValue(cell) - getValue(n) <= 1);
        return BFS.execute(this.end, isEnd, adjacent);
    }
    
    @Override
    public Integer solvePart1() {
        return solve(Set.of('S'));
    }

    @Override
    public Integer solvePart2() {
        return solve(Set.of('S', 'a'));
    }

    public static void main(final String[] args) throws Exception {
        assert AoC2022_12.createDebug(TEST).solvePart1() == 31;
        assert AoC2022_12.createDebug(TEST).solvePart2() == 29;

        final Puzzle puzzle = Aocd.puzzle(2022, 12);
        final List<String> inputData = puzzle.getInputData();
        puzzle.check(
            () -> lap("Part 1", AoC2022_12.create(inputData)::solvePart1),
            () -> lap("Part 2", AoC2022_12.create(inputData)::solvePart2)
        );
    }

    private static final List<String> TEST = splitLines("""
            Sabqponm
            abcryxxl
            accszExk
            acctuvwj
            abdefghi
            """);
}

File: src/main/python/AoC2022_12.py
#! /usr/bin/env python3
#
# Advent of Code 2022 Day 12
#

import sys

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples
from aoc.graph import bfs
from aoc.grid import Cell
from aoc.grid import CharGrid

TEST = """\
Sabqponm
abcryxxl
accszExk
acctuvwj
abdefghi
"""


Input = CharGrid
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return CharGrid.from_strings(list(input_data))

    def solve(self, grid: CharGrid, end_points: set[str]) -> int:
        def get_value(cell: Cell) -> int:
            ch = grid.get_value(cell)
            return (
                ord("a") if ch == "S" else ord("z") if ch == "E" else ord(ch)
            )

        distance, _ = bfs(
            next(grid.get_all_equal_to("E")),
            lambda cell: grid.get_value(cell) in end_points,
            lambda cell: (
                n
                for n in grid.get_capital_neighbours(cell)
                if get_value(cell) - get_value(n) <= 1
            ),
        )
        return distance

    def part_1(self, grid: Input) -> Output1:
        return self.solve(grid, {"S"})

    def part_2(self, grid: Input) -> Output2:
        return self.solve(grid, {"S", "a"})

    @aoc_samples(
        (
            ("part_1", TEST, 31),
            ("part_2", TEST, 29),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2022, 12)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()
